src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoOfT.cs

```diff
@@ -27,24 +27,84 @@ internal sealed class JsonPropertyInfo<T> : JsonPropertyInfo
         // the property's type, we track that and whether the property type can be null.
         private bool _propertyTypeEqualsTypeToConvert;
 
-        internal Func<object, T>? Get { get; set; }
+        private Func<object, T>? _typedGet;
+        private Action<object, T>? _typedSet;
 
-        internal Action<object, T>? Set { get; set; }
+        internal new Func<object, T>? Get
+        {
+            get => _typedGet;
+            set => SetGetter(value);
+        }
+
+        internal new Action<object, T>? Set
+        {
+            get => _typedSet;
+            set => SetSetter(value);
+        }
+
+        private protected override void SetGetter(Delegate? getter)
+        {
+            Debug.Assert(getter is null or Func<object, object?> or Func<object, T>);
+
+            CheckMutable();
+
+            if (getter is null)
+            {
+                _typedGet = null;
+                _untypedGet = null;
+            }
+            else if (getter is Func<object, T> typedGetter)
+            {
+                _typedGet = typedGetter;
+                _untypedGet = getter is Func<object, object?> untypedGet ? untypedGet : obj => typedGetter(obj);
+            }
+            else
+            {
+                Func<object, object?> untypedGet = (Func<object, object?>)getter;
+                _typedGet = (obj => (T)untypedGet(obj)!);
+                _untypedGet = untypedGet;
+            }
+        }
+
+        private protected override void SetSetter(Delegate? setter)
+        {
+            Debug.Assert(setter is null or Action<object, object?> or Action<object, T>);
+
+            CheckMutable();
+
+            if (setter is null)
+            {
+                _typedSet = null;
+                _untypedSet = null;
+            }
+            else if (setter is Action<object, T> typedSetter)
+            {
+                _typedSet = typedSetter;
+                _untypedSet = setter is Action<object, object?> untypedSet ? untypedSet : (obj, value) => typedSetter(obj, (T)value!);
+            }
+            else
+            {
+                Action<object, object?> untypedSet = (Action<object, object?>)setter;
+                _typedSet = ((obj, value) => untypedSet(obj, (T)value!));
+                _untypedSet = untypedSet;
+            }
+        }
 
         internal override object? DefaultValue => default(T);
 
-        public JsonConverter<T> Converter { get; internal set; } = null!;
+        internal JsonConverter<T> TypedEffectiveConverter { get; set; } = null!;
```

Comment:
```suggestion
        internal new JsonConverter<T> EffectiveConverter { get; set; } = null!;
```

Reply:
There is already `EffectiveConverter` in `JsonPropertyInfo<T>` - it makes sure converter is castable and assigns to the typed one. I can make setter private here though